src/matrix_auth_observer.py

```diff
@@ -0,0 +1,157 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The Matrix Auth relation observer."""
+
+import logging
+from collections import namedtuple
+from pathlib import Path
+from typing import List, Optional
+
+import ops
+from charms.synapse.v0.matrix_auth import (
+    MatrixAuthProviderData,
+    MatrixAuthProvides,
+    MatrixAuthRequirerData,
+)
+from ops.framework import Object
+
+import synapse
+from charm_state import CharmBaseWithState, CharmState, inject_charm_state
+
+logger = logging.getLogger(__name__)
+
+
+class MatrixAuthObserver(Object):
+    """The Matrix Auth relation observer."""
+
+    def __init__(self, charm: CharmBaseWithState):
+        """Initialize the observer and register event handlers.
+
+        Args:
+            charm: The parent charm to attach the observer to.
+        """
+        super().__init__(charm, "matrix-auth-observer")
+        self._charm = charm
+        self.matrix_auth = MatrixAuthProvides(self._charm)
+        # matrix_auth_request_received conflicts with on defined by Redis...
+        self.framework.observe(
+            self._charm.on["matrix-auth"].relation_changed, self._on_matrix_auth_relation_changed
+        )
+        self.framework.observe(
+            self._charm.on["matrix-auth"].relation_departed, self._on_matrix_auth_relation_departed
+        )
+
+    def get_charm(self) -> CharmBaseWithState:
+        """Return the current charm.
+
+        Returns:
+           The current charm
+        """
+        return self._charm
+
+    def update_matrix_auth_integration(self, charm_state: CharmState) -> None:
+        """Update matrix auth integration relation data.
+
+        Args:
+            charm_state: The charm state.
+        """
+        for relation in list(self._charm.model.relations["matrix-auth"]):
```

Comment:
Out of curiosity, is `list(...)` to guard against the relations being None?

Reply:
Exactly :)